### PR TITLE
Layer implementation

### PR DESCRIPTION
**Pull request by alexlemaire**
PR to review layer implementation

Implemented:
- added layer to current function config file
- added layer folder with base structure & files
- attached layer to global config file
- added scripts for:
  1. installing dependencies for all layers
  2. adding a new layer to our API (does not attach himself to a given function, this has to be done manually for now)
- refactored a bit `add-lambda` (and `add-layer`) since they mostly used the same logic/methods so I generalized the methods and moved them into `tools/add/helpers`

### CODE REVIEW

**Comment by Holger-Will** [comment]
https://klimapartner.slack.com/archives/DJXTZL6DN/p1583827741005800

File: tools/install-layers-dependencies/install-layers-dependencies.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+const fs = require('fs')
+const childProc = require('child_process')
+const root = process.cwd()
+
+fs.readdirSync('layers').forEach(layer => {
+  process.chdir(`layers/${layer}/nodejs`)
+  childProc.execSync('npm i')
```
[the remark this answers is not in the document]
Comment (alexlemaire) [reply tools/install-layers-dependencies/install-layers-dependencies.js]: Done!

File: package.json [hunk 2]
```diff
@@ -5,10 +5,13 @@
   "main": "serverless.yml",
   "scripts": {
     "test": "echo \"No test script defined.\"",
+    "dev-server.install": "cd tools/dev-server && npm i",
     "dev-server.start": "pm2 start tools/dev-server/index.js --name dev-server --watch",
     "dev-server.stop": "pm2 stop dev-server && pm2 delete dev-server",
-    "add-lambda": "node ./tools/add-lambda/add-lambda.js",
-    "update-version": "node ./tools/update-version/update-version.js"
+    "add-lambda": "node ./tools/add/add-lambda/add-lambda.js",
+    "add-layer": "node ./tools/add/add-layer/add-layer.js",
```
[the remark this answers is not in the document]
Comment (alexlemaire) [reply package.json]: Continuing this topic in a single comment below to centralize

File: serverless.json [hunk 3]
```diff
@@ -24,17 +24,19 @@
...
       "name": "kaskadi-serverless-deployment-bucket"
-    },
-    "deploymentPrefix": "${self:service.name}"
+    }
   },
   "functions": {
     "HelloWorld": "${file(./lambdas/hello-world/serverless.json)}"
+  },
+  "layers": {
+    "MyLayer": "${file(./layers/my-layer/serverless.json)}"
```
[the remark this answers is not in the document]
Comment (alexlemaire) [reply serverless.json]: Continuing this topic in a single comment below to centralize

**Comment by alexlemaire** [comment]
> https://klimapartner.slack.com/archives/DJXTZL6DN/p1583827741005800
> are we good with one layer per API?

I forgot to reply to this question... I was thinking we would still keep specialized layers (let's say `es-layer`, `wemalo-layer`, `sendemail-layer`, etc.) and then attach them only to the function that uses them to make it easier to maintain and have only the necessary dependencies attached to their respective functions

For example in the context of the authentication API using passwordless flow, all functions would probably use `es-layer` to read/write data in ElasticSearch but then `/login` & `/register` endpoint may use `sendemail-layer` while `/logout` probably not

In that case I think keeping specialized layers make sense. But tell me what you think! :+1:

**Comment by Holger-Will** [comment]
now that we group our Lambdas into APIs, we will keep the overhead quite small. it basically comes down to size-overhead vs. code-complexity. Here i would definitely vote for less code-complexity.

Its a complete tool we don't have to maintain, plus the devserver is simpler, plus you don't have to manage dependencies for single Lambdas, but only per API...

that is a lot compared to the potential savings on memory...

**Comment by alexlemaire** [comment]
As discussed together at the office yesterday: changed this to use a single layer :+1:

File: serverless.json [hunk 4]
```diff
@@ -31,6 +31,12 @@
     "region": "eu-central-1",
     "deploymentBucket": {
       "name": "kaskadi-serverless-deployment-bucket"
+    },
+    "tags": {
+      "app": "your-app-name",
+      "service": "${self:service.name}",
+      "logical-unit": "api-logical-unit",
+      "type": "api, s3, sns, etc."
```
Comment (alexlemaire) [inline serverless.json]: I implemented tags at API level following [this issue](https://github.com/kaskadi/template-kaskadi-api/issues/11#issue-578768729).

- `app` is our encapsulating app
- `service` no need to modify manually, this is our service
- `logical-unit` is the logical unit for our API like `stocks`, `auth`, `webhooks`, etc.
- `type` is the type of lambda we have here. Normally only `api` but maybe some are triggered by something else also and we want to tag it so we can find it again later on

**Comment by Holger-Will** [comment]
looking good! this can land...

File: serverless.json (hunk 4 above)
Comment (Holger-Will) [reply serverless.json]: i'm not sure about the names but otherwise looks nice. as i have no better idea for naming, we go with this for now...

**Pull request merged**